## Librerias

In [685]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import style
import seaborn as sns
import urllib.request
from PIL import Image


## Data

In [686]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
train = pd.read_csv("C:/Users/javier.sdiaz/Desktop\BootCamp/DS102024/3-MachineLearning/EDA_2/train.csv")


## Data exploration

In [687]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    float64
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
dtypes: float64(2), int64(1), object(10)
memory usage: 92.8+ KB


In [688]:
train.head()

laptop_ID Company             Product            TypeName  Inches  \
0       1223    Dell       Inspiron 5567            Notebook    15.6   
1         78  Lenovo  IdeaPad 320-15IKBN            Notebook    15.6   
2       1267    Dell              XPS 13  2 in 1 Convertible    13.3   
3        161    Dell       Inspiron 5579  2 in 1 Convertible    15.6   
4        922      LG         Gram 14Z970           Ultrabook    14.0   

                            ScreenResolution                         Cpu  Ram  \
0                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
1                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
2           Quad HD+ / Touchscreen 3200x1800   Intel Core i5 7Y54 1.2GHz  8GB   
3            Full HD / Touchscreen 1920x1080  Intel Core i7 8550U 1.8GHz  8GB   
4  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   

      Memory                     Gpu       OpSys  Weight  Price_euros  
0  256GB SSD      AMD Radeon R7 M445  Windows 10  2.36kg        889.0  
1    2TB HDD   Intel HD Graphics 620       No OS   2.2kg        519.0  
2  256GB SSD   Intel HD Graphics 615  Windows 10  1.24kg       1813.0  
3  256GB SSD  Intel UHD Graphics 620  Windows 10  1.56kg       1049.0  
4  512GB SSD   Intel HD Graphics 620  Windows 10  0.98kg       1899.0

In [689]:
import re

def separate_screen_resolution(resolution_str):
    # Extract resolution
    resolution_match = re.search(r'(\d+)\s*x\s*(\d+)', str(resolution_str))
    resolution = f"{resolution_match.group(1)}x{resolution_match.group(2)}" if resolution_match else None
    
    # Extract screen type
    screen_types = ['Touchscreen', 'IPS Panel', 'Full HD', 'Quad HD+', '4K Ultra HD', 'Retina Display']
    found_types = [type for type in screen_types if type.lower() in resolution_str.lower()]
    screen_type = ' / '.join(found_types) if found_types else None
    
    return screen_type, resolution

# Apply the function
train[['ScreenType', 'Resolution']] = train['ScreenResolution'].apply(separate_screen_resolution).apply(pd.Series)

In [690]:
# Crear una nueva columna 'IsTouchscreen' con True si contiene 'Touchscreen' y False si no
train['IsTouchscreen'] = train['ScreenType'].str.contains('Touchscreen', case=False, na=False)

In [691]:
train.head()

laptop_ID Company             Product            TypeName  Inches  \
0       1223    Dell       Inspiron 5567            Notebook    15.6   
1         78  Lenovo  IdeaPad 320-15IKBN            Notebook    15.6   
2       1267    Dell              XPS 13  2 in 1 Convertible    13.3   
3        161    Dell       Inspiron 5579  2 in 1 Convertible    15.6   
4        922      LG         Gram 14Z970           Ultrabook    14.0   

                            ScreenResolution                         Cpu  Ram  \
0                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
1                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
2           Quad HD+ / Touchscreen 3200x1800   Intel Core i5 7Y54 1.2GHz  8GB   
3            Full HD / Touchscreen 1920x1080  Intel Core i7 8550U 1.8GHz  8GB   
4  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   

      Memory                     Gpu       OpSys  Weight  Price_euros  \
0  256GB SSD      AMD Radeon R7 M445  Windows 10  2.36kg        889.0   
1    2TB HDD   Intel HD Graphics 620       No OS   2.2kg        519.0   
2  256GB SSD   Intel HD Graphics 615  Windows 10  1.24kg       1813.0   
3  256GB SSD  Intel UHD Graphics 620  Windows 10  1.56kg       1049.0   
4  512GB SSD   Intel HD Graphics 620  Windows 10  0.98kg       1899.0   

                          ScreenType Resolution  IsTouchscreen  
0                            Full HD  1920x1080          False  
1                            Full HD  1920x1080          False  
2             Touchscreen / Quad HD+  3200x1800           True  
3              Touchscreen / Full HD  1920x1080           True  
4  Touchscreen / IPS Panel / Full HD  1920x1080           True

In [692]:
# Separar el tipo de CPU y la memoria (en GHz)
train[['Cpu_Type', 'Memory_Cpu']] = train['Cpu'].str.extract(r'^(.*)\s(\d+\.?\d*GHz)$')

train.head()

laptop_ID Company             Product            TypeName  Inches  \
0       1223    Dell       Inspiron 5567            Notebook    15.6   
1         78  Lenovo  IdeaPad 320-15IKBN            Notebook    15.6   
2       1267    Dell              XPS 13  2 in 1 Convertible    13.3   
3        161    Dell       Inspiron 5579  2 in 1 Convertible    15.6   
4        922      LG         Gram 14Z970           Ultrabook    14.0   

                            ScreenResolution                         Cpu  Ram  \
0                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
1                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
2           Quad HD+ / Touchscreen 3200x1800   Intel Core i5 7Y54 1.2GHz  8GB   
3            Full HD / Touchscreen 1920x1080  Intel Core i7 8550U 1.8GHz  8GB   
4  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   

      Memory                     Gpu       OpSys  Weight  Price_euros  \
0  256GB SSD      AMD Radeon R7 M445  Windows 10  2.36kg        889.0   
1    2TB HDD   Intel HD Graphics 620       No OS   2.2kg        519.0   
2  256GB SSD   Intel HD Graphics 615  Windows 10  1.24kg       1813.0   
3  256GB SSD  Intel UHD Graphics 620  Windows 10  1.56kg       1049.0   
4  512GB SSD   Intel HD Graphics 620  Windows 10  0.98kg       1899.0   

                          ScreenType Resolution  IsTouchscreen  \
0                            Full HD  1920x1080          False   
1                            Full HD  1920x1080          False   
2             Touchscreen / Quad HD+  3200x1800           True   
3              Touchscreen / Full HD  1920x1080           True   
4  Touchscreen / IPS Panel / Full HD  1920x1080           True   

              Cpu_Type Memory_Cpu  
0  Intel Core i5 7200U     2.5GHz  
1  Intel Core i5 7200U     2.5GHz  
2   Intel Core i5 7Y54     1.2GHz  
3  Intel Core i7 8550U     1.8GHz  
4  Intel Core i7 7500U     2.7GHz

In [693]:
# Función para limpiar los valores y normalizarlos
def clean_and_normalize_memory(value):
    # Asegurarnos de que el valor es un string
    value = str(value).lower().strip()
    
    # Usar una expresión regular para capturar el número y la unidad opcional
    match = re.match(r'(\d+(\.\d+)?)(\s*(gb|mb|tb)?)?', value)
    
    if match:
        # Extraer el número y la unidad
        number = float(match.group(1))  # Captura el número
        unit = match.group(3).strip() if match.group(3) else ''  # Captura la unidad (puede estar vacía)
        
        # Convertir todo a GB
        if 'gb' in unit or unit == '':
            return number  # Ya está en GB o es solo un número
        elif 'mb' in unit:
            return number / 1024  # Convertir de MB a GB
        elif 'tb' in unit:
            return number * 1024  # Convertir de TB a GB
    return None  # Si no es válido, devolver None

# Aplicar la función a la columna Memory
train['Memory'] = train['Memory'].apply(clean_and_normalize_memory)

In [694]:
train.head()

laptop_ID Company             Product            TypeName  Inches  \
0       1223    Dell       Inspiron 5567            Notebook    15.6   
1         78  Lenovo  IdeaPad 320-15IKBN            Notebook    15.6   
2       1267    Dell              XPS 13  2 in 1 Convertible    13.3   
3        161    Dell       Inspiron 5579  2 in 1 Convertible    15.6   
4        922      LG         Gram 14Z970           Ultrabook    14.0   

                            ScreenResolution                         Cpu  Ram  \
0                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
1                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
2           Quad HD+ / Touchscreen 3200x1800   Intel Core i5 7Y54 1.2GHz  8GB   
3            Full HD / Touchscreen 1920x1080  Intel Core i7 8550U 1.8GHz  8GB   
4  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   

   Memory                     Gpu       OpSys  Weight  Price_euros  \
0   256.0      AMD Radeon R7 M445  Windows 10  2.36kg        889.0   
1  2048.0   Intel HD Graphics 620       No OS   2.2kg        519.0   
2   256.0   Intel HD Graphics 615  Windows 10  1.24kg       1813.0   
3   256.0  Intel UHD Graphics 620  Windows 10  1.56kg       1049.0   
4   512.0   Intel HD Graphics 620  Windows 10  0.98kg       1899.0   

                          ScreenType Resolution  IsTouchscreen  \
0                            Full HD  1920x1080          False   
1                            Full HD  1920x1080          False   
2             Touchscreen / Quad HD+  3200x1800           True   
3              Touchscreen / Full HD  1920x1080           True   
4  Touchscreen / IPS Panel / Full HD  1920x1080           True   

              Cpu_Type Memory_Cpu  
0  Intel Core i5 7200U     2.5GHz  
1  Intel Core i5 7200U     2.5GHz  
2   Intel Core i5 7Y54     1.2GHz  
3  Intel Core i7 8550U     1.8GHz  
4  Intel Core i7 7500U     2.7GHz

In [695]:
train[['OpsysType', 'Version']] = train['OpSys'].str.extract(r'^(Windows|Chrome OS|Linux|No OS|Mac OS X|macOS|Android)(?:\s+(.*))?$')

train.head()

laptop_ID Company             Product            TypeName  Inches  \
0       1223    Dell       Inspiron 5567            Notebook    15.6   
1         78  Lenovo  IdeaPad 320-15IKBN            Notebook    15.6   
2       1267    Dell              XPS 13  2 in 1 Convertible    13.3   
3        161    Dell       Inspiron 5579  2 in 1 Convertible    15.6   
4        922      LG         Gram 14Z970           Ultrabook    14.0   

                            ScreenResolution                         Cpu  Ram  \
0                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
1                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
2           Quad HD+ / Touchscreen 3200x1800   Intel Core i5 7Y54 1.2GHz  8GB   
3            Full HD / Touchscreen 1920x1080  Intel Core i7 8550U 1.8GHz  8GB   
4  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   

   Memory                     Gpu       OpSys  Weight  Price_euros  \
0   256.0      AMD Radeon R7 M445  Windows 10  2.36kg        889.0   
1  2048.0   Intel HD Graphics 620       No OS   2.2kg        519.0   
2   256.0   Intel HD Graphics 615  Windows 10  1.24kg       1813.0   
3   256.0  Intel UHD Graphics 620  Windows 10  1.56kg       1049.0   
4   512.0   Intel HD Graphics 620  Windows 10  0.98kg       1899.0   

                          ScreenType Resolution  IsTouchscreen  \
0                            Full HD  1920x1080          False   
1                            Full HD  1920x1080          False   
2             Touchscreen / Quad HD+  3200x1800           True   
3              Touchscreen / Full HD  1920x1080           True   
4  Touchscreen / IPS Panel / Full HD  1920x1080           True   

              Cpu_Type Memory_Cpu OpsysType Version  
0  Intel Core i5 7200U     2.5GHz   Windows      10  
1  Intel Core i5 7200U     2.5GHz     No OS     NaN  
2   Intel Core i5 7Y54     1.2GHz   Windows      10  
3  Intel Core i7 8550U     1.8GHz   Windows      10  
4  Intel Core i7 7500U     2.7GHz   Windows      10

In [696]:
train[['Manufacturer', 'MemoryGraphic']] = train['Gpu'].str.extract(r'^(Intel|AMD|Nvidia|ARM)\s+([\w\s]+)')

train.head()

laptop_ID Company             Product            TypeName  Inches  \
0       1223    Dell       Inspiron 5567            Notebook    15.6   
1         78  Lenovo  IdeaPad 320-15IKBN            Notebook    15.6   
2       1267    Dell              XPS 13  2 in 1 Convertible    13.3   
3        161    Dell       Inspiron 5579  2 in 1 Convertible    15.6   
4        922      LG         Gram 14Z970           Ultrabook    14.0   

                            ScreenResolution                         Cpu  Ram  \
0                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
1                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
2           Quad HD+ / Touchscreen 3200x1800   Intel Core i5 7Y54 1.2GHz  8GB   
3            Full HD / Touchscreen 1920x1080  Intel Core i7 8550U 1.8GHz  8GB   
4  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   

   Memory                     Gpu  ... Price_euros  \
0   256.0      AMD Radeon R7 M445  ...       889.0   
1  2048.0   Intel HD Graphics 620  ...       519.0   
2   256.0   Intel HD Graphics 615  ...      1813.0   
3   256.0  Intel UHD Graphics 620  ...      1049.0   
4   512.0   Intel HD Graphics 620  ...      1899.0   

                          ScreenType  Resolution IsTouchscreen  \
0                            Full HD   1920x1080         False   
1                            Full HD   1920x1080         False   
2             Touchscreen / Quad HD+   3200x1800          True   
3              Touchscreen / Full HD   1920x1080          True   
4  Touchscreen / IPS Panel / Full HD   1920x1080          True   

              Cpu_Type  Memory_Cpu OpsysType Version Manufacturer  \
0  Intel Core i5 7200U      2.5GHz   Windows      10          AMD   
1  Intel Core i5 7200U      2.5GHz     No OS     NaN        Intel   
2   Intel Core i5 7Y54      1.2GHz   Windows      10        Intel   
3  Intel Core i7 8550U      1.8GHz   Windows      10        Intel   
4  Intel Core i7 7500U      2.7GHz   Windows      10        Intel   

      MemoryGraphic  
0    Radeon R7 M445  
1   HD Graphics 620  
2   HD Graphics 615  
3  UHD Graphics 620  
4   HD Graphics 620  

[5 rows x 22 columns]

In [697]:
train['Model Base'] = train['Product'].str.extract(r'^([^(/]+)')  # Extrae el texto antes de un paréntesis o diagonal
train['Specifications'] = train['Product'].str.extract(r'\((.*)\)|/(.*)').bfill(axis=1)[0]  # Extrae el contenido entre paréntesis o después de una diagonal

train.head()

laptop_ID Company             Product            TypeName  Inches  \
0       1223    Dell       Inspiron 5567            Notebook    15.6   
1         78  Lenovo  IdeaPad 320-15IKBN            Notebook    15.6   
2       1267    Dell              XPS 13  2 in 1 Convertible    13.3   
3        161    Dell       Inspiron 5579  2 in 1 Convertible    15.6   
4        922      LG         Gram 14Z970           Ultrabook    14.0   

                            ScreenResolution                         Cpu  Ram  \
0                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
1                          Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   
2           Quad HD+ / Touchscreen 3200x1800   Intel Core i5 7Y54 1.2GHz  8GB   
3            Full HD / Touchscreen 1920x1080  Intel Core i7 8550U 1.8GHz  8GB   
4  IPS Panel Full HD / Touchscreen 1920x1080  Intel Core i7 7500U 2.7GHz  8GB   

   Memory                     Gpu  ... Resolution IsTouchscreen  \
0   256.0      AMD Radeon R7 M445  ...  1920x1080         False   
1  2048.0   Intel HD Graphics 620  ...  1920x1080         False   
2   256.0   Intel HD Graphics 615  ...  3200x1800          True   
3   256.0  Intel UHD Graphics 620  ...  1920x1080          True   
4   512.0   Intel HD Graphics 620  ...  1920x1080          True   

              Cpu_Type Memory_Cpu OpsysType  Version Manufacturer  \
0  Intel Core i5 7200U     2.5GHz   Windows       10          AMD   
1  Intel Core i5 7200U     2.5GHz     No OS      NaN        Intel   
2   Intel Core i5 7Y54     1.2GHz   Windows       10        Intel   
3  Intel Core i7 8550U     1.8GHz   Windows       10        Intel   
4  Intel Core i7 7500U     2.7GHz   Windows       10        Intel   

      MemoryGraphic          Model Base Specifications  
0    Radeon R7 M445       Inspiron 5567            NaN  
1   HD Graphics 620  IdeaPad 320-15IKBN            NaN  
2   HD Graphics 615              XPS 13            NaN  
3  UHD Graphics 620       Inspiron 5579            NaN  
4   HD Graphics 620         Gram 14Z970            NaN  

[5 rows x 24 columns]

In [698]:
train['Ram'] = train['Ram'].astype(str)
train['Ram'] = train['Ram'].str.replace('GB', '').astype(float)
train['Memory'] = train['Memory'].astype(str)
train['Memory'] = train['Memory'].str.replace('GHz', '').astype(float)
train['Weight']= train['Weight'].astype(str)
train['Weight'] = train['Weight'].str.replace('kg', '').astype(float)
train['Memory_Cpu']= train['Memory_Cpu'].astype(str)
train['Memory_Cpu'] = train['Memory_Cpu'].str.replace('GHz', '').astype(float)

In [699]:
train[['width', 'height']] = train['Resolution'].str.split('x', expand=True)

train['width'] = train['width'].astype(int)
train['height'] = train['height'].astype(int)

In [700]:
train.columns

Index(['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches',
       'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight',
       'Price_euros', 'ScreenType', 'Resolution', 'IsTouchscreen', 'Cpu_Type',
       'Memory_Cpu', 'OpsysType', 'Version', 'Manufacturer', 'MemoryGraphic',
       'Model Base', 'Specifications', 'width', 'height'],
      dtype='object')

In [701]:
df_filtered = train.drop(columns=['ScreenResolution', 'Cpu','Gpu','OpSys','Product'])
df_filtered

laptop_ID Company            TypeName  Inches   Ram  Memory  Weight  \
0         1223    Dell            Notebook    15.6   8.0   256.0    2.36   
1           78  Lenovo            Notebook    15.6   8.0  2048.0    2.20   
2         1267    Dell  2 in 1 Convertible    13.3   8.0   256.0    1.24   
3          161    Dell  2 in 1 Convertible    15.6   8.0   256.0    1.56   
4          922      LG           Ultrabook    14.0   8.0   512.0    0.98   
..         ...     ...                 ...     ...   ...     ...     ...   
907        709  Lenovo            Notebook    15.6   4.0   256.0    2.20   
908        947  Lenovo  2 in 1 Convertible    13.3  16.0  1024.0    1.30   
909        590    Dell            Notebook    15.6  16.0   256.0    1.93   
910       1143  Lenovo              Gaming    15.6   8.0   128.0    2.40   
911        531    Dell            Notebook    17.3   8.0   128.0    2.80   

     Price_euros                          ScreenType Resolution  ...  \
0         889.00                             Full HD  1920x1080  ...   
1         519.00                             Full HD  1920x1080  ...   
2        1813.00              Touchscreen / Quad HD+  3200x1800  ...   
3        1049.00               Touchscreen / Full HD  1920x1080  ...   
4        1899.00   Touchscreen / IPS Panel / Full HD  1920x1080  ...   
..           ...                                 ...        ...  ...   
907       399.00                             Full HD  1920x1080  ...   
908      1799.00  Touchscreen / IPS Panel / Quad HD+  3200x1800  ...   
909      1537.39                             Full HD  1920x1080  ...   
910       989.00                 IPS Panel / Full HD  1920x1080  ...   
911       889.00                 IPS Panel / Full HD  1920x1080  ...   

                 Cpu_Type Memory_Cpu  OpsysType Version Manufacturer  \
0     Intel Core i5 7200U        2.5    Windows      10          AMD   
1     Intel Core i5 7200U        2.5      No OS     NaN        Intel   
2      Intel Core i5 7Y54        1.2    Windows      10        Intel   
3     Intel Core i7 8550U        1.8    Windows      10        Intel   
4     Intel Core i7 7500U        2.7    Windows      10        Intel   
..                    ...        ...        ...     ...          ...   
907    AMD A9-Series 9420        2.9    Windows      10          AMD   
908   Intel Core i7 6560U        2.2    Windows      10        Intel   
909   Intel Core i7 7600U        2.8    Windows      10       Nvidia   
910  Intel Core i5 7300HQ        2.5      No OS     NaN       Nvidia   
911   Intel Core i5 8250U        1.6      Linux     NaN          AMD   

         MemoryGraphic          Model Base Specifications width  height  
0       Radeon R7 M445       Inspiron 5567            NaN  1920    1080  
1      HD Graphics 620  IdeaPad 320-15IKBN            NaN  1920    1080  
2      HD Graphics 615              XPS 13            NaN  3200    1800  
3     UHD Graphics 620       Inspiron 5579            NaN  1920    1080  
4      HD Graphics 620         Gram 14Z970            NaN  1920    1080  
..                 ...                 ...            ...   ...     ...  
907         Radeon 530   IdeaPad 320-15AST            NaN  1920    1080  
908  Iris Graphics 540      Yoga 900-13ISK            NaN  3200    1800  
909      GeForce 930MX       Latitude 5580            NaN  1920    1080  
910   GeForce GTX 1060  Legion Y520-15IKBN            NaN  1920    1080  
911         Radeon 530       Inspiron 5770            NaN  1920    1080  

[912 rows x 21 columns]

In [702]:
# Convertir la columna 'laptop_ID' a índice
df_filtered = train.set_index('laptop_ID')

In [703]:
# pasar la columna Manufacturer a numerica
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
df_filtered['Manufacturer'] = encoder.fit_transform(df_filtered['Manufacturer'])

In [704]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
df_filtered['OpsysType'] = encoder.fit_transform(df_filtered['OpsysType'])

In [705]:
encoder = LabelEncoder()
df_filtered['TypeName'] = encoder.fit_transform(df_filtered['TypeName'])

In [706]:
encoder = LabelEncoder()
df_filtered['Company'] = encoder.fit_transform(df_filtered['Company'])

In [707]:
df_filtered.columns

Index(['Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu',
       'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price_euros', 'ScreenType',
       'Resolution', 'IsTouchscreen', 'Cpu_Type', 'Memory_Cpu', 'OpsysType',
       'Version', 'Manufacturer', 'MemoryGraphic', 'Model Base',
       'Specifications', 'width', 'height'],
      dtype='object')

In [708]:
df_filtered = df_filtered.drop(columns=['IsTouchscreen','Memory','Weight','OpsysType','TypeName','Memory_Cpu','Product','ScreenResolution','Cpu','OpSys','Cpu_Type','MemoryGraphic','Model Base','Specifications','ScreenType','Resolution','Version','Gpu','width','height'])

In [710]:
#df_filtered['IsTouchscreen'] = df_filtered['IsTouchscreen'].map({True: 1, False: 0})

In [711]:
df_filtered.head()

Company  Inches  Ram  Price_euros  Manufacturer
laptop_ID                                                 
1223             4    15.6  8.0        889.0             0
78              10    15.6  8.0        519.0             1
1267             4    13.3  8.0       1813.0             1
161              4    15.6  8.0       1049.0             1
922              9    14.0  8.0       1899.0             1

In [712]:
df_filtered.to_csv('train_normalizado.csv', index=False, sep=';', encoding='utf-8')

## Data processing

In [713]:
train = df_filtered

In [714]:
# df.to_csv("Dataset/limpio.csv")

-----------------------------------------------------------------------------------------------------------------

## Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [715]:
X = train.drop(['Price_euros'], axis=1)
y = train['Price_euros'].copy()
X.shape

(912, 4)

In [716]:
y.shape

(912,)

### 2. Dividir X_train, X_test, y_train, y_test

In [717]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [718]:
X_train

Company  Inches  Ram  Manufacturer
laptop_ID                                    
439             10    14.0  4.0             1
979              4    15.6  4.0             0
337             16    15.6  8.0             1
139             10    15.6  4.0             1
517              4    13.3  8.0             1
...            ...     ...  ...           ...
474              4    15.6  8.0             0
478              7    15.6  8.0             0
952              4    14.0  8.0             1
370             10    15.6  8.0             2
1226             0    17.3  8.0             2

[729 rows x 4 columns]

### 3. Asignar el modelo (vacío) a una variable


In [719]:
from sklearn.linear_model import LinearRegression

regressor = LinearRegression()

regressor.fit(X_train, y_train)

LinearRegression()

### 4. Sacar métricas, valorar el modelo 

Recuerda que en la competición se va a evaluar con la métrica de RMSE.

In [720]:
from sklearn.metrics import root_mean_squared_error

predictions = regressor.predict(X_test)

rmse = root_mean_squared_error(y_test, predictions)

In [721]:
print(rmse)

486.0733786801218


-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `test.csv` nos arriesgamos a que digamos, no salga lo esperado.

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `train.csv` al completo.


**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**


Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `test.csv` para predecir.

In [325]:
X_pred = pd.read_csv("C:/Users/javier.sdiaz/Desktop/BootCamp/DS102024/3-MachineLearning/EDA_2/test.csv")
X_pred.head()

laptop_ID   Company                 Product   TypeName  Inches  \
0        539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1        327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2        563  Mediacom           SmartBook 130   Notebook    13.3   
3         13     Apple             MacBook Pro  Ultrabook    15.4   
4        935        HP           EliteBook 850  Ultrabook    15.6   

                     ScreenResolution                          Cpu   Ram  \
0     IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                   Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2         IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3  IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                   Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   

                 Memory                      Gpu       OpSys  Weight  
0  256GB SSD +  1TB HDD  Nvidia GeForce GTX 960M  Windows 10     2kg  
1             256GB SSD    Intel HD Graphics 620  Windows 10     2kg  
2    32GB Flash Storage        Intel HD Graphics  Windows 10  1.35kg  
3             256GB SSD       AMD Radeon Pro 555       macOS  1.83kg  
4             256GB SSD      AMD Radeon R7 M365X  Windows 10  1.84kg

In [326]:
X_pred.shape

(391, 12)

In [327]:
X_pred

laptop_ID   Company                 Product   TypeName  Inches  \
0          539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1          327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2          563  Mediacom           SmartBook 130   Notebook    13.3   
3           13     Apple             MacBook Pro  Ultrabook    15.4   
4          935        HP           EliteBook 850  Ultrabook    15.6   
..         ...       ...                     ...        ...     ...   
386        742    Lenovo             ThinkPad 13   Notebook    13.3   
387        660      Dell                  XPS 13  Ultrabook    13.3   
388        983    Lenovo       IdeaPad 310-15IKB   Notebook    15.6   
389       1137        HP          EliteBook 1040   Notebook    14.0   
390        556    Lenovo           ThinkPad T570   Notebook    15.6   

                       ScreenResolution                          Cpu   Ram  \
0       IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                     Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2           IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3    IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                     Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   
..                                  ...                          ...   ...   
386         IPS Panel Full HD 1920x1080   Intel Core i5 7200U 2.5GHz   8GB   
387                   Full HD 1920x1080   Intel Core i5 8250U 1.6GHz   8GB   
388                   Full HD 1920x1080   Intel Core i5 7200U 2.5GHz   6GB   
389                   Full HD 1920x1080   Intel Core i5 6200U 2.3GHz   8GB   
390         IPS Panel Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   

                   Memory                      Gpu       OpSys  Weight  
0    256GB SSD +  1TB HDD  Nvidia GeForce GTX 960M  Windows 10     2kg  
1               256GB SSD    Intel HD Graphics 620  Windows 10     2kg  
2      32GB Flash Storage        Intel HD Graphics  Windows 10  1.35kg  
3               256GB SSD       AMD Radeon Pro 555       macOS  1.83kg  
4               256GB SSD      AMD Radeon R7 M365X  Windows 10  1.84kg  
..                    ...                      ...         ...     ...  
386             256GB SSD    Intel HD Graphics 620  Windows 10  1.44kg  
387             256GB SSD   Intel UHD Graphics 620  Windows 10   1.2kg  
388             256GB SSD     Nvidia GeForce 920MX  Windows 10   2.4kg  
389             256GB SSD    Intel HD Graphics 520   Windows 7  1.43kg  
390             512GB SSD    Intel HD Graphics 620  Windows 10  1.99kg  

[391 rows x 12 columns]

In [328]:
import re

def separate_screen_resolution(resolution_str):
    # Extract resolution
    resolution_match = re.search(r'(\d+)\s*x\s*(\d+)', str(resolution_str))
    resolution = f"{resolution_match.group(1)}x{resolution_match.group(2)}" if resolution_match else None
    
    # Extract screen type
    screen_types = ['Touchscreen', 'IPS Panel', 'Full HD', 'Quad HD+', '4K Ultra HD', 'Retina Display']
    found_types = [type for type in screen_types if type.lower() in resolution_str.lower()]
    screen_type = ' / '.join(found_types) if found_types else None
    
    return screen_type, resolution


X_pred[['ScreenType', 'Resolution']] = X_pred['ScreenResolution'].apply(separate_screen_resolution).apply(pd.Series)

In [329]:
# Crear una nueva columna 'IsTouchscreen' con True si contiene 'Touchscreen' y False si no
X_pred['IsTouchscreen'] = X_pred['ScreenType'].str.contains('Touchscreen', case=False, na=False)

# Separar el tipo de CPU y la memoria (en GHz)
X_pred[['Cpu_Type', 'Memory_Cpu']] = X_pred['Cpu'].str.extract(r'^(.*)\s(\d+\.?\d*GHz)$')



In [330]:
# Función para limpiar los valores y normalizarlos
def clean_and_normalize_memory(value):
    # Asegurarnos de que el valor es un string
    value = str(value).lower().strip()
    
    # Usar una expresión regular para capturar el número y la unidad opcional
    match = re.match(r'(\d+(\.\d+)?)(\s*(gb|mb|tb)?)?', value)
    
    if match:
        # Extraer el número y la unidad
        number = float(match.group(1))  # Captura el número
        unit = match.group(3).strip() if match.group(3) else ''  # Captura la unidad (puede estar vacía)
        
        # Convertir todo a GB
        if 'gb' in unit or unit == '':
            return number  # Ya está en GB o es solo un número
        elif 'mb' in unit:
            return number / 1024  # Convertir de MB a GB
        elif 'tb' in unit:
            return number * 1024  # Convertir de TB a GB
    return None  # Si no es válido, devolver None

In [331]:
# Aplicar la función a la columna Memory
X_pred['Memory'] = X_pred['Memory'].apply(clean_and_normalize_memory)

X_pred[['OpsysType', 'Version']] = X_pred['OpSys'].str.extract(r'^(Windows|Chrome OS|Linux|No OS|Mac OS X|macOS|Android)(?:\s+(.*))?$')



X_pred[['Manufacturer', 'MemoryGraphic']] = X_pred['Gpu'].str.extract(r'^(Intel|AMD|Nvidia|ARM)\s+([\w\s]+)')



X_pred['Model Base'] = X_pred['Product'].str.extract(r'^([^(/]+)')  # Extrae el texto antes de un paréntesis o diagonal
X_pred['Specifications'] = X_pred['Product'].str.extract(r'\((.*)\)|/(.*)').bfill(axis=1)[0]  # Extrae el contenido entre paréntesis o después de una diagonal



X_pred['Ram'] = X_pred['Ram'].astype(str)
X_pred['Ram'] = X_pred['Ram'].str.replace('GB', '').astype(float)
X_pred['Memory'] = X_pred['Memory'].astype(str)
X_pred['Memory'] = X_pred['Memory'].str.replace('GHz', '').astype(float)
X_pred['Weight']= X_pred['Weight'].astype(str)
X_pred['Weight'] = X_pred['Weight'].str.replace('kg', '').astype(float)
X_pred['Memory_Cpu']= X_pred['Memory_Cpu'].astype(str)
X_pred['Memory_Cpu'] = X_pred['Memory_Cpu'].str.replace('GHz', '').astype(float)

X_pred[['width', 'height']] = X_pred['Resolution'].str.split('x', expand=True)

X_pred['width'] = X_pred['width'].astype(int)
X_pred['height'] = X_pred['height'].astype(int)

In [332]:
df_test = X_pred.drop(columns=['ScreenResolution', 'Cpu','Gpu','OpSys','Product'])
df_test

laptop_ID   Company   TypeName  Inches   Ram  Memory  Weight  \
0          539      Asus   Notebook    15.6   8.0   256.0    2.00   
1          327      Asus   Notebook    14.0   8.0   256.0    2.00   
2          563  Mediacom   Notebook    13.3   4.0    32.0    1.35   
3           13     Apple  Ultrabook    15.4  16.0   256.0    1.83   
4          935        HP  Ultrabook    15.6   8.0   256.0    1.84   
..         ...       ...        ...     ...   ...     ...     ...   
386        742    Lenovo   Notebook    13.3   8.0   256.0    1.44   
387        660      Dell  Ultrabook    13.3   8.0   256.0    1.20   
388        983    Lenovo   Notebook    15.6   6.0   256.0    2.40   
389       1137        HP   Notebook    14.0   8.0   256.0    1.43   
390        556    Lenovo   Notebook    15.6   8.0   512.0    1.99   

                     ScreenType Resolution  IsTouchscreen  \
0       IPS Panel / 4K Ultra HD  3840x2160          False   
1                       Full HD  1920x1080          False   
2           IPS Panel / Full HD  1920x1080          False   
3    IPS Panel / Retina Display  2880x1800          False   
4                       Full HD  1920x1080          False   
..                          ...        ...            ...   
386         IPS Panel / Full HD  1920x1080          False   
387                     Full HD  1920x1080          False   
388                     Full HD  1920x1080          False   
389                     Full HD  1920x1080          False   
390         IPS Panel / Full HD  1920x1080          False   

                Cpu_Type  Memory_Cpu OpsysType Version Manufacturer  \
0    Intel Core i7 7500U        2.70   Windows      10       Nvidia   
1    Intel Core i7 7500U        2.70   Windows      10        Intel   
2    Intel Atom x5-Z8350        1.44   Windows      10        Intel   
3          Intel Core i7        2.80     macOS     NaN          AMD   
4    Intel Core i7 6500U        2.50   Windows      10          AMD   
..                   ...         ...       ...     ...          ...   
386  Intel Core i5 7200U        2.50   Windows      10        Intel   
387  Intel Core i5 8250U        1.60   Windows      10        Intel   
388  Intel Core i5 7200U        2.50   Windows      10       Nvidia   
389  Intel Core i5 6200U        2.30   Windows       7        Intel   
390  Intel Core i7 7500U        2.70   Windows      10        Intel   

        MemoryGraphic              Model Base Specifications  width  height  
0    GeForce GTX 960M  Zenbook UX510UW-FI095T            NaN   3840    2160  
1     HD Graphics 620  ZenBook UX410UA-GV183T            NaN   1920    1080  
2         HD Graphics           SmartBook 130            NaN   1920    1080  
3      Radeon Pro 555             MacBook Pro            NaN   2880    1800  
4     Radeon R7 M365X           EliteBook 850            NaN   1920    1080  
..                ...                     ...            ...    ...     ...  
386   HD Graphics 620             ThinkPad 13            NaN   1920    1080  
387  UHD Graphics 620                  XPS 13            NaN   1920    1080  
388     GeForce 920MX       IdeaPad 310-15IKB            NaN   1920    1080  
389   HD Graphics 520          EliteBook 1040            NaN   1920    1080  
390   HD Graphics 620           ThinkPad T570            NaN   1920    1080  

[391 rows x 20 columns]

In [333]:
# Convertir la columna 'laptop_ID' a índice
df_test = df_test.set_index('laptop_ID')


In [334]:
df_test

Company   TypeName  Inches   Ram  Memory  Weight  \
laptop_ID                                                      
539            Asus   Notebook    15.6   8.0   256.0    2.00   
327            Asus   Notebook    14.0   8.0   256.0    2.00   
563        Mediacom   Notebook    13.3   4.0    32.0    1.35   
13            Apple  Ultrabook    15.4  16.0   256.0    1.83   
935              HP  Ultrabook    15.6   8.0   256.0    1.84   
...             ...        ...     ...   ...     ...     ...   
742          Lenovo   Notebook    13.3   8.0   256.0    1.44   
660            Dell  Ultrabook    13.3   8.0   256.0    1.20   
983          Lenovo   Notebook    15.6   6.0   256.0    2.40   
1137             HP   Notebook    14.0   8.0   256.0    1.43   
556          Lenovo   Notebook    15.6   8.0   512.0    1.99   

                           ScreenType Resolution  IsTouchscreen  \
laptop_ID                                                         
539           IPS Panel / 4K Ultra HD  3840x2160          False   
327                           Full HD  1920x1080          False   
563               IPS Panel / Full HD  1920x1080          False   
13         IPS Panel / Retina Display  2880x1800          False   
935                           Full HD  1920x1080          False   
...                               ...        ...            ...   
742               IPS Panel / Full HD  1920x1080          False   
660                           Full HD  1920x1080          False   
983                           Full HD  1920x1080          False   
1137                          Full HD  1920x1080          False   
556               IPS Panel / Full HD  1920x1080          False   

                      Cpu_Type  Memory_Cpu OpsysType Version Manufacturer  \
laptop_ID                                                                   
539        Intel Core i7 7500U        2.70   Windows      10       Nvidia   
327        Intel Core i7 7500U        2.70   Windows      10        Intel   
563        Intel Atom x5-Z8350        1.44   Windows      10        Intel   
13               Intel Core i7        2.80     macOS     NaN          AMD   
935        Intel Core i7 6500U        2.50   Windows      10          AMD   
...                        ...         ...       ...     ...          ...   
742        Intel Core i5 7200U        2.50   Windows      10        Intel   
660        Intel Core i5 8250U        1.60   Windows      10        Intel   
983        Intel Core i5 7200U        2.50   Windows      10       Nvidia   
1137       Intel Core i5 6200U        2.30   Windows       7        Intel   
556        Intel Core i7 7500U        2.70   Windows      10        Intel   

              MemoryGraphic              Model Base Specifications  width  \
laptop_ID                                                                   
539        GeForce GTX 960M  Zenbook UX510UW-FI095T            NaN   3840   
327         HD Graphics 620  ZenBook UX410UA-GV183T            NaN   1920   
563             HD Graphics           SmartBook 130            NaN   1920   
13           Radeon Pro 555             MacBook Pro            NaN   2880   
935         Radeon R7 M365X           EliteBook 850            NaN   1920   
...                     ...                     ...            ...    ...   
742         HD Graphics 620             ThinkPad 13            NaN   1920   
660        UHD Graphics 620                  XPS 13            NaN   1920   
983           GeForce 920MX       IdeaPad 310-15IKB            NaN   1920   
1137        HD Graphics 520          EliteBook 1040            NaN   1920   
556         HD Graphics 620           ThinkPad T570            NaN   1920   

           height  
laptop_ID          
539          2160  
327          1080  
563          1080  
13           1800  
935          1080  
...           ...  
742          1080  
660          1080  
983          1080  
1137         1080  
556          1080  

[391 rows x 19 columns]

In [335]:
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
df_test['Manufacturer'] = encoder.fit_transform(df_test['Manufacturer'])

from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
df_test['OpsysType'] = encoder.fit_transform(df_test['OpsysType'])

encoder = LabelEncoder()
df_test['TypeName'] = encoder.fit_transform(df_test['TypeName'])

encoder = LabelEncoder()
df_test['Company'] = encoder.fit_transform(df_test['Company'])

In [336]:
df_test.columns

Index(['Company', 'TypeName', 'Inches', 'Ram', 'Memory', 'Weight',
       'ScreenType', 'Resolution', 'IsTouchscreen', 'Cpu_Type', 'Memory_Cpu',
       'OpsysType', 'Version', 'Manufacturer', 'MemoryGraphic', 'Model Base',
       'Specifications', 'width', 'height'],
      dtype='object')

In [337]:
df_test = df_test.drop(columns=['Cpu_Type','MemoryGraphic','Model Base','Specifications','ScreenType','Resolution','Version','width','height'])

df_test['IsTouchscreen'] = df_test['IsTouchscreen'].map({True: 1, False: 0})


In [338]:
df_test.to_csv('test_normalizado.csv', index=False, sep=';', encoding='utf-8')

**¿Por qué da error?** 

In [339]:
predictions_submit = regressor.predict(df_test)
predictions_submit

array([1360.15727443, 1378.19733507,  844.92697037, 1789.16027966,
       1037.14060604,  567.85128553, 2051.51793695,  506.70430426,
        958.03963025,  872.18546373, 1217.20952496, 1908.0784425 ,
        332.239426  ,  773.09855872, 1268.82123522, 2158.65975694,
        342.23050337, 1011.83281995, 2044.3521048 ,  483.60794188,
       1114.70780948, 1365.81624574, 1940.82996803,  598.85508157,
        823.2326295 , 1051.12490969,  891.17117008, 1990.78481382,
       1372.88925354,  535.88292184, 1746.38970013, 1489.40085238,
        307.53538099, 2137.79959988, 1442.17575926,  427.16206739,
       1268.82123522, 1826.11729895,  663.36772128, 1269.35188946,
       1854.18673187, 1409.9281291 , 1858.52625752, 1231.72411006,
       1286.21297332,  830.22121501,  708.92397391, 3514.28280481,
        861.12215403, 1089.65546033,  710.10196956,  756.57082962,
        452.40199247,  531.26316035, 1159.5440111 ,  306.66955537,
       1216.84614506, 1251.33703604, 1922.09615825, 1087.02143

IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 391 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `test.csv`?**

In [340]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [288]:
X_pred = pd.read_csv("test.csv")
X_pred.head()

laptop_ID   Company                 Product   TypeName  Inches  \
0        539      Asus  Zenbook UX510UW-FI095T   Notebook    15.6   
1        327      Asus  ZenBook UX410UA-GV183T   Notebook    14.0   
2        563  Mediacom           SmartBook 130   Notebook    13.3   
3         13     Apple             MacBook Pro  Ultrabook    15.4   
4        935        HP           EliteBook 850  Ultrabook    15.6   

                     ScreenResolution                          Cpu   Ram  \
0     IPS Panel 4K Ultra HD 3840x2160   Intel Core i7 7500U 2.7GHz   8GB   
1                   Full HD 1920x1080   Intel Core i7 7500U 2.7GHz   8GB   
2         IPS Panel Full HD 1920x1080  Intel Atom x5-Z8350 1.44GHz   4GB   
3  IPS Panel Retina Display 2880x1800         Intel Core i7 2.8GHz  16GB   
4                   Full HD 1920x1080   Intel Core i7 6500U 2.5GHz   8GB   

                 Memory                      Gpu       OpSys  Weight  
0  256GB SSD +  1TB HDD  Nvidia GeForce GTX 960M  Windows 10     2kg  
1             256GB SSD    Intel HD Graphics 620  Windows 10     2kg  
2    32GB Flash Storage        Intel HD Graphics  Windows 10  1.35kg  
3             256GB SSD       AMD Radeon Pro 555       macOS  1.83kg  
4             256GB SSD      AMD Radeon R7 M365X  Windows 10  1.84kg

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [341]:
predictions_submit = regressor.predict(df_test)
predictions_submit

array([1360.15727443, 1378.19733507,  844.92697037, 1789.16027966,
       1037.14060604,  567.85128553, 2051.51793695,  506.70430426,
        958.03963025,  872.18546373, 1217.20952496, 1908.0784425 ,
        332.239426  ,  773.09855872, 1268.82123522, 2158.65975694,
        342.23050337, 1011.83281995, 2044.3521048 ,  483.60794188,
       1114.70780948, 1365.81624574, 1940.82996803,  598.85508157,
        823.2326295 , 1051.12490969,  891.17117008, 1990.78481382,
       1372.88925354,  535.88292184, 1746.38970013, 1489.40085238,
        307.53538099, 2137.79959988, 1442.17575926,  427.16206739,
       1268.82123522, 1826.11729895,  663.36772128, 1269.35188946,
       1854.18673187, 1409.9281291 , 1858.52625752, 1231.72411006,
       1286.21297332,  830.22121501,  708.92397391, 3514.28280481,
        861.12215403, 1089.65546033,  710.10196956,  756.57082962,
        452.40199247,  531.26316035, 1159.5440111 ,  306.66955537,
       1216.84614506, 1251.33703604, 1922.09615825, 1087.02143

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [342]:
sample = pd.read_csv("sample_submission.csv")

In [343]:
sample.head()

laptop_ID  Price_euros
0        539          650
1        327          650
2        563          650
3         13          650
4        935          500

In [344]:
sample.shape

(391, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [345]:
submission = pd.DataFrame({"laptop_ID": sample['laptop_ID'], "Price_euros": predictions_submit})

In [346]:
submission.head()

laptop_ID  Price_euros
0        539  1360.157274
1        327  1378.197335
2        563   844.926970
3         13  1789.160280
4        935  1037.140606

In [347]:
submission.shape

(391, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [348]:
def chequeator(df_to_submit,version):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit['laptop_ID'].all() == sample['laptop_ID'].all():
                print("You're ready to submit!")
                submission.to_csv(f"submission_{version}.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [349]:
chequeator(submission,1)

You're ready to submit!
